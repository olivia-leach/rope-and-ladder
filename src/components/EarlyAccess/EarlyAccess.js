import React, { useState, useEffect, useCallback } from 'react'
import Loader from '../Loader/Loader';

import './EarlyAccess.scss'

function EarlyAccess(props) {
    const [isValid, setIsValid] = useState(true)
    const [isLoading, setIsLoading] = useState(true)
    
    useEffect(() => {
        // lock body scroll
        document.body.classList.add('isFixed')

        // check for token
        const urlHasCode = props.location.search.includes('?code=')
        const codeIs10 = props.location.search.length === 16
        if (!urlHasCode || !codeIs10) {
            setIsValid(false)
            setIsLoading(false)
        }

        return function cleanup() {
            document.body.classList.remove('isFixed')
        }
    }, [props.location.search])

    const goHome = useCallback(() => {
        props.history.push('/')
    }, [props.history])

    const stopProp = useCallback((e) => {
        e.stopPropagation()
    }, [])

    const setLoadingFalse = useCallback(() => {
        setIsLoading(false)
    }, [])

    const setLoadingTrue = useCallback(() => {
        setIsLoading(true)
    }, [])

    return (
        <div className='EarlyAccess' onClick={goHome}>
            <div className="EarlyAccess__Modal" onClick={stopProp}>
                <button className="EarlyAccess__Modal__Button" onClick={goHome}>
                    <span aria='hidden'>x</span>
                    <span className='visuallyhidden'>Close</span>
                </button>
                <div className={`EarlyAccess__Modal__Body${isLoading ? '--isLoading' : ''}`}>
                    {isLoading && <div className='EarlyAccess__LoaderContainer'><Loader /></div>}
                    {!isLoading && isValid && (
                        <div>
                            <h2>Thanks for signing up.</h2>
                            <p>Enjoy your early access.</p>
                            <p>So far, we've released two singles, but â€” This project was never meant to be listened to one single at a time. It's a concept album: The songs tell a cohesive story in chronological order.</p>
                            <p>The first four songs of the record make up the prologue - A little bit of storytelling that happens a few years before tracks 5-13.</p>
                            <p>We'd love for you to set aside some time and approach it as you would film or theatre. Let your imagination fill out the scenes, and start getting to know the characters.</p>
                            <p>Let us know what you think! Shoot us a message - we're always excited to chat.</p>
                            <h4>Love,<br/>Rope and Ladder</h4>
                        </div>
                    )}
                    {!isLoading && !isValid && (
                        <div>
                            <h2>Uh oh, you forgot something.</h2>
                            <p>Sign up on the home page to get your personal early access link.</p>
                            <h4>Love,<br/>Rope and Ladder</h4>
                        </div>
                    )}
                    {isValid && (
                        <iframe
                            // style={{ border: 0, width: '350px', height: '786px' }}
                            style={{ border: 0, width: '400px', height: '472px' }}
                            // src="https://bandcamp.com/EmbeddedPlayer/album=2731032401/size=large/bgcol=ffffff/linkcol=e99708/artwork=small/transparent=true/tracks=2496975933,1440423378,1973265551,91971970,437267700"
                            // src="https://bandcamp.com/EmbeddedPlayer/album=2731032401/size=large/bgcol=ffffff/linkcol=0687f5/transparent=true/tracklist=true/tracks=2496975933,1440423378,1973265551,91971970,437267700"
                            // src="https://bandcamp.com/EmbeddedPlayer/album=2731032401/size=large/bgcol=ffffff/linkcol=0687f5/transparent=true/tracklist=true/tracks=2496975933,1440423378,1973265551,91971970,437267700/esig=a1560812820944b3236fd301b66c580a"
                            src="https://bandcamp.com/EmbeddedPlayer/album=2731032401/size=large/bgcol=ffffff/linkcol=e99708/artwork=none/transparent=true"
                            seamless
                            title='bandcamp'
                            onLoad={setLoadingFalse}
                            onError={setLoadingTrue}
                        >
                        <a href="http://ropeandladder.bandcamp.com/album/rope-and-ladder-2">Rope and Ladder by Rope and Ladder</a>
                    </iframe>
                    )}
                </div>
            </div>
        </div>
    )
}

export default EarlyAccess