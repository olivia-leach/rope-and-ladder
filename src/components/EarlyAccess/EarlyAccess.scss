.EarlyAccess {
    width: 100%;
    height: 100%;
    position: fixed;
    overflow-y: scroll;
    top: 0;
    left: 0;
    z-index: 2147483646;
    background: rgba(0, 0, 0, 0.6);

    &__Modal {
        margin: 100px auto 40px auto;
        border-radius: 2px;
        border: 1px solid rgba(0, 0, 0, 0.72);
        max-width: 1000px;
        background: #fff;
        width: calc(100% - 20px);
        position: absolute;
        left: 0;
        right: 0;
        z-index: 2147483646;

        &__Button {
            top: -55px;
            color: #fff;
            right: 0;
            z-index: 2147483647;
            padding: 0 10px;
            font-weight: 100;
            font-size: 40px;
            position: absolute;
            background-color: transparent;
            border: none;
            transition: transform 100ms ease, color 100ms ease, -webkit-transform 100ms ease;
            cursor: pointer;
            font-family: "Lato", Helvetica, Arial, sans-serif;

            &:hover {
                transform: scale(1.2);
            }
        }

        &__Body {
            padding: 2em;
            color: #4c4c4c;
            display: flex;
            justify-content: space-between;

            &--isLoading {
                iframe {
                    opacity: 0;
                }
            }
        }

        iframe {
            flex-shrink: 0;
            margin: 0 0 0 4em;
        }
    }

    &__LoaderContainer {
        width: 100%;
        height: 100%;
        display: flex;
        justify-content: center;
        align-items: center;
        position: absolute;
        left: 0;
        top: 0;
    }

    .lds-roller {
        position: absolute;
        margin: 0 auto;
    }
}

.visuallyhidden {
    border: 0;
    height: 1px;
    margin: -1px;
    overflow: hidden;
    padding: 0;
    position: absolute;
    width: 1px;
}