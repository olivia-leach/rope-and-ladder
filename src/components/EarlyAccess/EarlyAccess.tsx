import React, { useState, useEffect, useCallback } from 'react'
import Loader from '../Loader/Loader'
import { useHistory, useLocation } from 'react-router'

import './EarlyAccess.scss'

const stopProp = (e: React.MouseEvent<HTMLDivElement>) => e.stopPropagation()

const EarlyAccess = () => {
  const [isValid, setIsValid] = useState(true)
  const [isLoading, setIsLoading] = useState(true)
  const { push } = useHistory()
  const location = useLocation()

  useEffect(() => {
    // lock body scroll
    document.body.classList.add('isFixed')

    // check for token
    const urlHasCode = location.search.includes('?code=')
    const codeIs10 = location.search.length === 16
    if (!urlHasCode || !codeIs10) {
      setIsValid(false)
      setIsLoading(false)
    }

    return function cleanup() {
      document.body.classList.remove('isFixed')
    }
  }, [location.search])

  const goHome = useCallback(() => {
    push('/')
  }, [push])

  const setLoadingFalse = useCallback(() => {
    setIsLoading(false)
  }, [])

  return (
    <div className="EarlyAccess" onClick={goHome}>
      <div className="EarlyAccess__Modal" onClick={stopProp}>
        <button className="EarlyAccess__Modal__Button" onClick={goHome} data-testid='closeButton'>
          <span aria-hidden="true">x</span>
          <span className="visuallyhidden">Close</span>
        </button>
        <div
          className={`EarlyAccess__Modal__Body${
            isLoading ? '--isLoading' : ''
          }`}
        >
          {isLoading && (
            <div className="EarlyAccess__LoaderContainer">
              <Loader />
            </div>
          )}
          {!isLoading && isValid && (
            <div>
              <h2>Thanks for signing up.</h2>
              <p>Enjoy your early access.</p>
              <p>
                So far, we've released two singles, but â€” This project was never
                meant to be listened to one single at a time. It's a concept
                album: The songs tell a cohesive story in chronological order.
              </p>
              <p>
                The first four songs of the record make up the prologue - A
                little bit of storytelling that happens a few years before
                tracks 5-13.
              </p>
              <p>
                We'd love for you to set aside some time and approach it as you
                would film or theatre. Let your imagination fill out the scenes,
                and start getting to know the characters.
              </p>
              <p>
                Let us know what you think! Shoot us a message - we're always
                excited to chat.
              </p>
              <h4>
                Love,
                <br />
                Rope and Ladder
              </h4>
            </div>
          )}
          {!isLoading && !isValid && (
            <div>
              <h2>Uh oh, you forgot something.</h2>
              <p>
                Sign up on the home page to get your personal early access link.
              </p>
              <h4>
                Love,
                <br />
                Rope and Ladder
              </h4>
            </div>
          )}
          {isValid && (
            <iframe
              data-testid='bandCampIFrame'
              style={{ border: 0, width: '400px', height: '472px' }}
              src="https://bandcamp.com/EmbeddedPlayer/album=2731032401/size=large/bgcol=ffffff/linkcol=e99708/artwork=none/transparent=true"
              seamless
              title="bandcamp"
              onLoad={setLoadingFalse}
            >
              <a href="http://ropeandladder.bandcamp.com/album/rope-and-ladder-2">
                Rope and Ladder by Rope and Ladder
              </a>
            </iframe>
          )}
        </div>
      </div>
    </div>
  )
}

export default EarlyAccess
