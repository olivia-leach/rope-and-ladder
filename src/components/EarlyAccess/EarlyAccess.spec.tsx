import { render, cleanup, fireEvent } from '@testing-library/react'
import React from 'react'
import { StaticRouter } from 'react-router-dom'
import EarlyAccess from './EarlyAccess'

afterEach(cleanup)

const mockHistoryPush = jest.fn()

jest.mock('react-router', () => ({
  ...(jest.requireActual('react-router') as object),
  useHistory: () => ({
    push: mockHistoryPush,
  }),
}))

describe('EarlyAccess', () => {
  it('should match snapshot if access code is valid', () => {
    const { container } = render(
      <StaticRouter location={{ search: 'code=1234567890' }}>
        <EarlyAccess />
      </StaticRouter>
    )
    expect(container).toMatchSnapshot()
  })

  it('should match snapshot if access code is invalid', () => {
    const { container } = render(
      <StaticRouter location={{ search: 'code=123' }}>
        <EarlyAccess />
      </StaticRouter>
    )
    expect(container).toMatchSnapshot()
  })

  it('should navigate home when close button is clicked', () => {
    const { getByTestId } = render(
      <StaticRouter location={{ search: 'code=1234567890' }}>
        <EarlyAccess />
      </StaticRouter>
    )
  
    fireEvent.click(getByTestId('closeButton'))
    expect(mockHistoryPush).toHaveBeenCalledTimes(1)
    expect(mockHistoryPush).toHaveBeenCalledWith('/')
  })

  it('should match snapshot when iframe loads', () => {
    const { container, getByTestId } = render(
      <StaticRouter location={{ search: 'code=1234567890' }}>
        <EarlyAccess />
      </StaticRouter>
    )
    const iframe = getByTestId('bandCampIFrame')
    fireEvent.load(iframe)
    expect(container).toMatchSnapshot()
  })
})
