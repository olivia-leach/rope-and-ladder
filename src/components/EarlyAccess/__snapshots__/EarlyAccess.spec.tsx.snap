// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EarlyAccess should match snapshot if access code is invalid 1`] = `
<div>
  <div
    class="EarlyAccess"
  >
    <div
      class="EarlyAccess__Modal"
    >
      <button
        class="EarlyAccess__Modal__Button"
        data-testid="closeButton"
      >
        <span
          aria-hidden="true"
        >
          x
        </span>
        <span
          class="visuallyhidden"
        >
          Close
        </span>
      </button>
      <div
        class="EarlyAccess__Modal__Body"
      >
        <div>
          <h2>
            Uh oh, you forgot something.
          </h2>
          <p>
            Sign up on the home page to get your personal early access link.
          </p>
          <h4>
            Love,
            <br />
            Rope and Ladder
          </h4>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`EarlyAccess should match snapshot if access code is valid 1`] = `
<div>
  <div
    class="EarlyAccess"
  >
    <div
      class="EarlyAccess__Modal"
    >
      <button
        class="EarlyAccess__Modal__Button"
        data-testid="closeButton"
      >
        <span
          aria-hidden="true"
        >
          x
        </span>
        <span
          class="visuallyhidden"
        >
          Close
        </span>
      </button>
      <div
        class="EarlyAccess__Modal__Body--isLoading"
      >
        <div
          class="EarlyAccess__LoaderContainer"
        >
          <div
            class="lds-roller"
          >
            <div />
            <div />
            <div />
            <div />
            <div />
            <div />
            <div />
            <div />
          </div>
        </div>
        <iframe
          data-testid="bandCampIFrame"
          seamless=""
          src="https://bandcamp.com/EmbeddedPlayer/album=2731032401/size=large/bgcol=ffffff/linkcol=e99708/artwork=none/transparent=true"
          style="border: 0px; width: 400px; height: 472px;"
          title="bandcamp"
        >
          <a
            href="http://ropeandladder.bandcamp.com/album/rope-and-ladder-2"
          >
            Rope and Ladder by Rope and Ladder
          </a>
        </iframe>
      </div>
    </div>
  </div>
</div>
`;

exports[`EarlyAccess should match snapshot when iframe loads 1`] = `
<div>
  <div
    class="EarlyAccess"
  >
    <div
      class="EarlyAccess__Modal"
    >
      <button
        class="EarlyAccess__Modal__Button"
        data-testid="closeButton"
      >
        <span
          aria-hidden="true"
        >
          x
        </span>
        <span
          class="visuallyhidden"
        >
          Close
        </span>
      </button>
      <div
        class="EarlyAccess__Modal__Body"
      >
        <div>
          <h2>
            Thanks for signing up.
          </h2>
          <p>
            Enjoy your early access.
          </p>
          <p>
            So far, we've released two singles, but â€” This project was never meant to be listened to one single at a time. It's a concept album: The songs tell a cohesive story in chronological order.
          </p>
          <p>
            The first four songs of the record make up the prologue - A little bit of storytelling that happens a few years before tracks 5-13.
          </p>
          <p>
            We'd love for you to set aside some time and approach it as you would film or theatre. Let your imagination fill out the scenes, and start getting to know the characters.
          </p>
          <p>
            Let us know what you think! Shoot us a message - we're always excited to chat.
          </p>
          <h4>
            Love,
            <br />
            Rope and Ladder
          </h4>
        </div>
        <iframe
          data-testid="bandCampIFrame"
          seamless=""
          src="https://bandcamp.com/EmbeddedPlayer/album=2731032401/size=large/bgcol=ffffff/linkcol=e99708/artwork=none/transparent=true"
          style="border: 0px; width: 400px; height: 472px;"
          title="bandcamp"
        >
          <a
            href="http://ropeandladder.bandcamp.com/album/rope-and-ladder-2"
          >
            Rope and Ladder by Rope and Ladder
          </a>
        </iframe>
      </div>
    </div>
  </div>
</div>
`;
